File: app/src/main/java/com/qimian233/ztool/hook/modules/mobiledesktop/AutoAcceptFileTransferHook.java
package com.qimian233.ztool.hook.modules.mobiledesktop;

import com.qimian233.ztool.data.keys.ScopeKeys;
import com.qimian233.ztool.hook.base.AppHookModule;

import io.github.libxposed.api.XposedModuleInterface;

import java.lang.reflect.Field;
import java.lang.reflect.Method;

/**
 * 自动接受超级互联 PC→手机 文件互传确认弹窗。
 * <p>
 * 在 FileConnectionConfirmActivity.onCreate 完成后直接通过 ViewModel
 * 触发接受逻辑，实现弹窗出现即自动确认，无需用户手动点击。
 * </p>
 * <p>
 * 使用 DEXKit 验证目标类存在，通过反射按<b>类型签名</b>查找 ViewModel 字段和
 * LiveData 字段，不再依赖混淆后的字段名（c/d/b）。
 * </p>
 */
public class AutoAcceptFileTransferHook extends AppHookModule {

    private static final String TARGET_PACKAGE = ScopeKeys.MOBILE_DESKTOP.packageName;
    private static final String TARGET_CLASS =
            "com.motorola.mobiledesktop.files.pc2phone.FileConnectionConfirmActivity";

    public AutoAcceptFileTransferHook() {}

    @Override
    public String getModuleName() {
        return "auto_accept_file_transfer";
    }

    @Override
    public String[] getTargetPackages() {
        return new String[]{TARGET_PACKAGE};
    }

    @Override
    public void handleLoadPackage(XposedModuleInterface.PackageLoadedParam param) throws Throwable {
        ClassLoader classLoader = param.getDefaultClassLoader();

        try {
            Class<?> activityClass = classLoader.loadClass(TARGET_CLASS);

            // 1. 在 Activity 中按类型查找 ViewModel 字段：检查类型是否继承 ViewModel
            String vmFieldName = findFieldBySuperType(activityClass
                    /* 回退名称 */);

            // 2. 从找到的字段获取 ViewModel 类，再在其中查找 boolean 和 LiveData 字段
            Class<?> vmClass;
            Field vmField;
            try {
                vmField = activityClass.getDeclaredField(vmFieldName);
                vmField.setAccessible(true);
                vmClass = vmField.getType();
            } catch (NoSuchFieldException e) {
                // 用回退名称 "c" 再试
                vmField = activityClass.getDeclaredField("c");
                vmField.setAccessible(true);
                vmClass = vmField.getType();
            }

            // 3. 在 ViewModel 中按类型查找 boolean 和 LiveData 字段
            String acceptedFieldName;
            String liveDataFieldName;
            acceptedFieldName = findFieldByType(vmClass
                    /* 回退 */);
            liveDataFieldName = findLiveDataField(vmClass  /* 回退 */);

            final String finalVmFieldName = vmFieldName;
            final String finalAcceptedFieldName = acceptedFieldName;
            final String finalLiveDataFieldName = liveDataFieldName;

            this.getXposed().hook(activityClass.getDeclaredMethod("onCreate",
                    android.os.Bundle.class)).intercept(chain -> {
                Object result = chain.proceed();

                Object activity = chain.getThisObject();
                try {
                    Field vmF = activityClass.getDeclaredField(finalVmFieldName);
                    vmF.setAccessible(true);
                    Object viewModel = vmF.get(activity);
                    if (viewModel == null) {
                        logger.warn("ViewModel is null, skip auto-accept");
                        return result;
                    }

                    Class<?> vmCls = viewModel.getClass();

                    Field acceptedF = vmCls.getDeclaredField(finalAcceptedFieldName);
                    acceptedF.setAccessible(true);
                    acceptedF.setBoolean(viewModel, true);

                    Field liveDataF = vmCls.getDeclaredField(finalLiveDataFieldName);
                    liveDataF.setAccessible(true);
                    Object liveData = liveDataF.get(viewModel);
                    if (liveData != null) {
                        Method updateMethod = findLiveDataUpdateMethod(liveData.getClass());
                        if (updateMethod != null) {
                            updateMethod.invoke(liveData, Boolean.TRUE);
                            logger.debug("Auto-accepted file transfer" +
                                    " [vm=" + finalVmFieldName +
                                    ", accepted=" + finalAcceptedFieldName +
                                    ", ld=" + finalLiveDataFieldName + "]");
                        } else {
                            logger.warn("Cannot find LiveData update method, skip");
                        }
                    } else {
                        logger.warn("LiveData field is null, skip");
                    }
                } catch (Throwable t) {
                    logger.error("Failed to auto-accept file transfer", t);
                }
                return result;
            });
            logger.info("Installed hook for auto-accept file transfer");
        } catch (Throwable t) {
            logger.error("Failed to install auto-accept file transfer hook", t);
        }
    }

    /**
     * 遍历类的所有字段，按类型查找目标字段。
     */
    private static String findFieldByType(Class<?> clazz) {
        for (Field f : clazz.getDeclaredFields()) {
            String typeName = f.getType().getName();
            if ("boolean".equals(typeName)
                    || f.getType().getName().endsWith("." + "boolean")) {
                return f.getName();
            }
        }
        return "d";
    }

    /**
     * 在类中查找其类型继承链包含指定超类的字段。
     */
    private static String findFieldBySuperType(Class<?> clazz) {
        for (Field f : clazz.getDeclaredFields()) {
            if (isSubclassOf(f.getType(), "androidx.lifecycle.ViewModel")) {
                return f.getName();
            }
        }
        return "c";
    }

    /**
     * 在类中查找 LiveData/MutableLiveData 类型的字段。
     * 通过检查字段类型的继承链是否包含 androidx.lifecycle.LiveData，
     * 而非依靠方法签名启发式匹配，避免匹配到同名签名的其他类。
     */
    private static String findLiveDataField(Class<?> clazz) {
        for (Field f : clazz.getDeclaredFields()) {
            if (isSubclassOf(f.getType(), "androidx.lifecycle.LiveData")) {
                return f.getName();
            }
        }
        // 回退：检查类型名包含 LiveData（针对未混淆的情况）
        for (Field f : clazz.getDeclaredFields()) {
            String tn = f.getType().getName();
            if (tn.contains("LiveData")) {
                return f.getName();
            }
        }
        return "b";
    }

    /**
     * 检查 cls 的继承链中是否包含指定名称的超类或接口。
     */
    private static boolean isSubclassOf(Class<?> cls, String superName) {
        while (cls != null && cls != Object.class) {
            if (superName.equals(cls.getName())) {
                return true;
            }
            // 也检查实现的接口
            for (Class<?> iface : cls.getInterfaces()) {
                if (superName.equals(iface.getName())) {
                    return true;
                }
            }
            cls = cls.getSuperclass();
        }
        return false;
    }

    /**
     * 在 LiveData/MutableLiveData 类层次中按参数签名查找更新方法。
     * 混淆后 setValue → l, postValue → i，二者签名均为 (Object)void。
     */
    private static Method findLiveDataUpdateMethod(Class<?> cls) {
        while (cls != null && cls != Object.class) {
            for (Method m : cls.getDeclaredMethods()) {
                Class<?>[] params = m.getParameterTypes();
                if (params.length == 1 && params[0] == Object.class
                        && m.getReturnType() == void.class) {
                    m.setAccessible(true);
                    return m;
                }
            }
            cls = cls.getSuperclass();
        }
        return null;
    }
}

File: app/src/main/java/com/qimian233/ztool/hook/modules/mobiledesktop/AutoAcceptFileTransferHook.kt
package com.qimian233.ztool.hook.modules.mobiledesktop

import android.os.Bundle
import com.qimian233.ztool.data.keys.PreferenceKeys
import com.qimian233.ztool.data.keys.ScopeKeys
import com.qimian233.ztool.dexindex.base.DexIndexConstants
import com.qimian233.ztool.hook.base.AppHookModule
import com.qimian233.ztool.hook.base.DexIndexStore
import io.github.libxposed.api.XposedModuleInterface.PackageLoadedParam
import java.lang.reflect.Field
import java.lang.reflect.Method

/**
 * Auto-accepts the PC→phone file transfer confirmation dialog of Super
 * Interconnect (FileUnion).
 * <p>
 * After FileConnectionConfirmActivity.onCreate completes, the accept logic is
 * triggered directly through the ViewModel, so the dialog is auto-confirmed as
 * soon as it appears, without requiring the user to tap manually.
 * </p>
 * <p>
 * Target class/field/method names are pre-computed via the DexKit offline
 * index (MobileDesktopDexIndexer); if the index is missing, hardcoded names
 * for the current version are used as fallback.
 * </p>
 */
class AutoAcceptFileTransferHook : AppHookModule() {

    override fun getModuleName(): String = PreferenceKeys.AUTO_ACCEPT_FILE_TRANSFER.name

    override fun getTargetPackages(): Array<String> = arrayOf(ScopeKeys.MOBILE_DESKTOP.packageName)

    override fun handleLoadPackage(param: PackageLoadedParam) {
        val classLoader = param.defaultClassLoader

        val vmFieldName = DexIndexStore.string(
            xposed, ScopeKeys.MOBILE_DESKTOP.packageName,
            DexIndexConstants.ModuleKeys.AUTO_ACCEPT_FILE_TRANSFER,
            DexIndexConstants.Keys.VM_FIELD_NAME
        ) ?: FALLBACK_VM_FIELD
        val acceptedFieldName = DexIndexStore.string(
            xposed, ScopeKeys.MOBILE_DESKTOP.packageName,
            DexIndexConstants.ModuleKeys.AUTO_ACCEPT_FILE_TRANSFER,
            DexIndexConstants.Keys.ACCEPTED_FIELD_NAME
        ) ?: FALLBACK_ACCEPTED_FIELD
        val liveDataFieldName = DexIndexStore.string(
            xposed, ScopeKeys.MOBILE_DESKTOP.packageName,
            DexIndexConstants.ModuleKeys.AUTO_ACCEPT_FILE_TRANSFER,
            DexIndexConstants.Keys.LIVE_DATA_FIELD_NAME
        ) ?: FALLBACK_LIVE_DATA_FIELD
        val liveDataUpdateMethodName = DexIndexStore.string(
            xposed, ScopeKeys.MOBILE_DESKTOP.packageName,
            DexIndexConstants.ModuleKeys.AUTO_ACCEPT_FILE_TRANSFER,
            DexIndexConstants.Keys.LIVE_DATA_UPDATE_METHOD
        )

        try {
            val activityClass = classLoader.loadClass(TARGET_CLASS)

            val vmField = try {
                activityClass.getDeclaredField(vmFieldName)
            } catch (_: NoSuchFieldException) {
                // Retry with fallback name "c"
                activityClass.getDeclaredField(FALLBACK_VM_FIELD)
            }
            vmField.isAccessible = true
            val vmClass = vmField.type

            val finalVmFieldName = vmField.name

            hookWithId(activityClass.getDeclaredMethod("onCreate", Bundle::class.java), "on_create") { chain ->
                val result = chain.proceed()
                val activity = chain.thisObject
                try {
                    val vmF: Field = activityClass.getDeclaredField(finalVmFieldName)
                    vmF.isAccessible = true
                    val viewModel = vmF.get(activity)
                    if (viewModel == null) {
                        logger.warn("ViewModel is null, skip auto-accept")
                        return@hookWithId result
                    }

                    val vmCls = viewModel.javaClass

                    val acceptedF = try {
                        vmCls.getDeclaredField(acceptedFieldName)
                    } catch (_: NoSuchFieldException) {
                        vmCls.getDeclaredField(FALLBACK_ACCEPTED_FIELD)
                    }
                    acceptedF.isAccessible = true
                    acceptedF.setBoolean(viewModel, true)

                    val liveDataF = try {
                        vmCls.getDeclaredField(liveDataFieldName)
                    } catch (_: NoSuchFieldException) {
                        vmCls.getDeclaredField(FALLBACK_LIVE_DATA_FIELD)
                    }
                    liveDataF.isAccessible = true
                    val liveData = liveDataF.get(viewModel)
                    if (liveData != null) {
                        val updateMethod = liveDataUpdateMethodName?.let { methodName ->
                            try {
                                liveData.javaClass.getDeclaredMethod(methodName, Any::class.java)
                            } catch (_: NoSuchMethodException) {
                                null
                            }
                        } ?: findLiveDataUpdateMethod(liveData.javaClass)
                        if (updateMethod != null) {
                            updateMethod.invoke(liveData, java.lang.Boolean.TRUE)
                            logger.debug(
                                "Auto-accepted file transfer [vm=$finalVmFieldName, " +
                                    "accepted=${acceptedF.name}, ld=${liveDataF.name}]"
                            )
                        } else {
                            logger.warn("Cannot find LiveData update method, skip")
                        }
                    } else {
                        logger.warn("LiveData field is null, skip")
                    }
                } catch (t: Throwable) {
                    logger.error("Failed to auto-accept file transfer", t)
                }
                result
            }
            logger.info("Installed hook for auto-accept file transfer")
        } catch (t: Throwable) {
            logger.error("Failed to install auto-accept file transfer hook", t)
        }
    }

    /**
     * Finds an update method by parameter signature in the LiveData/MutableLiveData
     * class hierarchy. After obfuscation setValue → l and postValue → i; both have
     * the signature (Object)void. Used only as a fallback when the offline index is
     * missing (values are pre-computed with DexKit at index time).
     */
    private fun findLiveDataUpdateMethod(cls: Class<*>): Method? {
        var current: Class<*>? = cls
        while (current != null && current != Any::class.java) {
            for (m in current.declaredMethods) {
                val params = m.parameterTypes
                if (params.size == 1 && params[0] == Any::class.java
                    && m.returnType == Void.TYPE
                ) {
                    m.isAccessible = true
                    return m
                }
            }
            current = current.superclass
        }
        return null
    }

    companion object {
        private const val TARGET_CLASS =
            "com.motorola.mobiledesktop.files.pc2phone.FileConnectionConfirmActivity"
        private const val FALLBACK_VM_FIELD = "a"
        private const val FALLBACK_ACCEPTED_FIELD = "d"
        private const val FALLBACK_LIVE_DATA_FIELD = "b"
    }
}
